{% extends "d3b_generic.html" %}

{% block bodytitle %}
{{ header }}
{% endblock %}

{% block bodycontent %}

<script>
//document.getElementById('upload_file').addEventListener('change', handleFileSelect, false);

function setuploadname( evt ) 
{
    var files = evt.target.files; 

	var dname = document.getElementById( "id_name" );
	if ( dname.value.length > 0 ) return;

	for (var i = 0, f; f = files[i]; i++) 
	{
		fname = files[i].name;
		dname.value = fname.substring( 0, fname.lastIndexOf('.') );
    }
}


</script>
<br><br>
<div class="sectionheader">
Submit new matrix 
</div>

<form action="/" method="post" enctype="multipart/form-data">
    {% csrf_token %} 
    {{ form.as_p }}
    <input type="submit" value="Submit" />
</form>
<script>
document.getElementById('id_file').addEventListener('change', setuploadname, false);
</script>

<br><br>
<div class="sectionheader">
Links to sample datasets 
</div>
<table>
<tr><td><a href="summary/00000001">marine sponges</a><td class="sampledescription"> [Roche 454, 15 samples, original study: <a href="http://dx.doi.org/10.1007/s12526-017-0697-0">doi:10.1007/s12526-017-0697-0</a>, QIIME1 open reference pipeline: <a href="http://dx.doi.org/10.1016/j.resmic.2018.03.004">doi:10.1016/j.resmic.2018.03.004</a>]
<tr><td><a href="summary/00000002">goose gut microbiome</a><td class="sampledescription"> [Illumina MiSeq, 9 samples, original study: <a href="https://doi.org/10.1016/j.micres.2015.10.003">doi:10.1016/j.micres.2015.10.003</a>, QIIME1 open reference pipeline: <a href="http://dx.doi.org/10.1016/j.resmic.2018.03.004">doi:10.1016/j.resmic.2018.03.004</a> ]
<tr><td><a href="summary/00000003">osteomyelitis in jaw bones</a><td class="sampledescription"> [Roche 454, 9 samples, oroginal study: <a href="https://doi.org/10.1111/1469-0691.12400">doi:10.1111/1469-0691.12400</a>, QIIME1 open reference pipeline: <a href="http://dx.doi.org/10.1016/j.resmic.2018.03.004">doi:10.1016/j.resmic.2018.03.004</a> ]
<tr><td><a href="summary/00000004">sponges in Baikal lake</a><td class="sampledescription"> [Roche 454, 23 samples, oroginal study: <a href="https://doi.org/10.1007/s00248-017-1097-5">doi:10.1007/s00248-017-1097-5</a>, QIIME1 closed reference pipeline: <a href="http://bri-shur.com/downloads/feranchuk-limnolfwbiol-fractalmodels.pdf">pdf</a>, doi:10.31951/2658-3518-2018-A-2-122 ]
<tr><td><a href="summary/00000005">counts of marine molluscs</a><td class="sampledescription"> [original data: <a href="http://www.iobis.org/">OBIS site</a>, custom pipeline: <a href="http://dx.doi.org/10.1016/j.resmic.2018.03.004">doi:10.1016/j.resmic.2018.03.004</a> ]
</table>
<br><br>
<div class="sectionheader">
Description 
</div>
<script>
function expand_description()
{
 document.getElementById( "description" ).style.display = "block";
}
</script>
[<span onclick="expand_description()" style="cursor:pointer"><i><b>click to expand</b></i>]
<div id="description" style="display:none">
<p>The online services available on this site have been developed for the extensive analysis of tables with estimates of bacteria abundance levels in environmental samples. In most cases the microbial ecology-specific functionality is implemented by the scikit-bio Python package, together with the other Python packages intended for analysis of big datasets. The interactive visualisation tools are implemented by the D3.js software library; therefore, the software project is named D3b. The source codes are available at github as an installable package.</p>
<h2>Installation</h2>
<p>The installation process may require the skills of a system administrator, if there are incompatibilities between the python3 environment and the software dependencies.</p>
<p>Suggested installation via conda:</p>

<p>- Create virtual environment:
<br><span class="monospace">conda create -n d3b python=3.6 anaconda</span>
<br><span class="monospace">source activate d3b</span>
</p>
<p>- Download the package and run setup.py:
<br><span class="monospace">git clone https://github.com/sferanchuk/d3b_charts.git</span>
<br><span class="monospace">cd d3b_charts/d3b</span>
<br><span class="monospace">pip install .</span>
</p>
<p>-  Download the phantomjs package, extract the executable file from the archive and move it to the 'phantomjs' subdirectory. Suggested URLs to obtain phantomjs:
<a href="https://bitbucket.org/ariya/phantomjs/downloads/phantomjs-2.1.1-windows.zip" rel="nofollow">Windows</a>;
<a href="https://bitbucket.org/ariya/phantomjs/downloads/phantomjs-2.1.1-linux-x86_64.tar.bz2" rel="nofollow">Linux x86_64</a>;
<a href="https://bitbucket.org/ariya/phantomjs/downloads/phantomjs-2.1.1-linux-i686.tar.bz2" rel="nofollow">Linux i686</a>;
<a href="https://bitbucket.org/ariya/phantomjs/downloads/phantomjs-2.1.1-macosx.zip" rel="nofollow">"Apple"</a>.

<br>This step is optional; phantomjs is used to render graphics on a server side and is required to export PNG and SVG images. Other packages external to the python environment, like d3.js, are already included in the distribution.</p>
<p>- Run the django server (runserver.sh) in console. Point the browser to URL 'http://localhost:8000'. Modify the ALLOWED_HOSTS variable in d3b/settings.py to resolve any problems with remote access to pages.</p>
<h2>Getting started</h2>
<p>Use links to the sample datasets on a start page to explore the abilities of the d3b system. On each link, the menu provides tools to analyse a dataset in various representations.</p><p>For each tool, the result becomes available in the browser after pressing the 'Submit' button. The input parameters should be specified before submitting the task. In order to become familiar with the design and conventions of the d3b system, it is advisable to first try out the default settings on the sample datasets,&nbsp;</p><p>For this purpose, datasets composed of abundances of annotated OTUs in microbial communities are provided. Four of the sample datasets are estimates of amplicon sequencing followed by the QIIME1 pipeline, and one dataset represents a selection from a custom ecological community.</p><p>All the datasets are represented as matrices of abundance counts. To allow the specification of traits and taxonomic groups, the descriptors for the rows and columns of the matrix can be modified and saved. The tools which allow to specify both kinds of descriptors (Subsets of samples / Taxonomic groups) are also available in the sidebar menu for any dataset.
<h2>Usage</h2>
<p>To apply the tools from the d3b system to custom data, a dataset in the form of a matrix should be loaded to the server side of the system. After that, a permanent link to the processed dataset becomes available. It should be kept and can be accessed at any time when the whole system is up and running.</p><p>Two formats of input matrix are accepted. The first is tab-delimited format (TSV) and the second is BIOM format. In both formats, abundance counts for OTU are combined with taxonomic annotation of the OTU and identifiers of the samples.</p><p>A tab-delimited file exported from a spreadsheet using packages like MS Excel or LibreOfiice Calc is suitable for the system, provided the conventions for the contents of rows and columns are followed. And the processed dataset can be exported back to tab-delimited format, as provided for in the tool 'matrix of abundances'.</p><p>The output obtained from that tool applied to one of the sample datasets can provide an example of a correctly composed tab-delimited file. In that file, each line begins from a complete taxonomic annotation for the OTU, in a fixed number of cells, followed by absolute counts of abundances for this OTU in all of the samples. In the first line, cells with numbers from 1 to 2, 3, etc., are used to specify fixed levels of hierarchy in the taxonomic system, and they are followed by cells with identifiers of the samples.</p><p>Data in BIOM format ("biological observation matrix") is also suitable for import into d3b. This format is often used to keep intermediate output results in software systems for microbiologists. But due to the flexibility of the BIOM format specification, the import from a particular BIOM-formatted file may fail, or may give wrong results. In this case, the python script 'biom2emap.py' from the source codes of d3b system can be used as a template for a user-defined script, to convert any biom-formatted dataset to a correct tab-delimited file.</p>

<h2>Overview of tools</h2>
<p>The tabular presentations of the data include a table of abundances, just as it is loaded into the system. There are various options for sorting the rows, merging the columns, reducing the level of taxonomic hierarchy, etc.. Most of the input forms include a choice of level of taxonomic hierarchy, the possibility to restrict the analysis or data presentation to certain taxonomic groups, and the possibility to merge samples into pre-defined groups.</p><div><br class="m_5627933771320707064webkit-block-placeholder"></div><p>The tabular presentations also include:&nbsp;<br>1) the values of alpha-diversity calculated using several of the most informative estimators,&nbsp;<br>2) the significance of differences for alpha-diversity values between several groups,&nbsp;<br>3) the significance of differences between groups of samples calculated using the distances between samples, with several alternative measures of distance.</p><p>The graphical presentations, implemented with the use of the d3.js library, currenly include the following charts:&nbsp;<br>1) A bubble chart and heatmap, to represent absolute/relative abundances.&nbsp;<br>2) 2D scatter charts, to represent the results of several data ordination methods, such as principal component analysis (PCA) or multi-dimensional scaling (MDS). The choice of several measures is available here to calculate distances between the samples.&nbsp;<br>3) A dendrogram (tree) to represent the degree of proximity between samples.&nbsp;<br>4) A Venn diagram to represent the unique and shared taxonomic units for the samples, implemented with the use of the jVenn plugin and the venn.js library.&nbsp;<br>5) Two kinds of diagrams to present distributions which describe a sample or a group of samples: a rank-abundance chart (Whittaker plot) to represent the distribution of relative species abundance, and a rarefaction curve to estimate the effect of insufficient coverage and sample size.&nbsp;<br>6) A ternary chart, to represent the relative abundances of bacterial phylotypes for the three samples, or groups of samples.&nbsp;<br>7) A Volcano chart and mean-distance plot, to represent distribution of abundances and differentiation between traits.&nbsp;<br>8) Two combined 2D charts, to represent the results of PCA decomposition applied directly to a non-square matrix of abundances. One chart is for the samples in the survey and the second adjacent chart is for the bacterial species in the rows of the submitted matrix.</p>
</div>
{% endblock %}


